atc/configvalidate/validate.go

```diff
@@ -490,3 +496,62 @@ func validateDisplay(c atc.Config) ([]atc.ConfigWarning, error) {
 
 	return warnings, nil
 }
+
+func detectCycle(j atc.JobConfig, visited map[string]int, initialJobConfig atc.Config) (bool, string) {
+	const (
+		nonVisited = 0
+		semiVisited = 1
+		alreadyVisited = 2
+	)
+	visited[j.Name] = semiVisited
+	state := false
+	cycleErrorJob := ""
+	_ = j.StepConfig().Visit(atc.StepRecursor{
+		OnGet: func(step *atc.GetStep) error {
+			for _, nextJobName := range step.Passed {
+				nextJob := findJobByName(nextJobName, initialJobConfig.Jobs)
+				if visited[nextJobName] == semiVisited {
+					state = true
+					cycleErrorJob = nextJobName
+					return nil
+				} else if visited[nextJobName] == nonVisited {
+					nextState, jobErrorName := detectCycle(nextJob, visited, initialJobConfig)
+					state = state || nextState
+					if jobErrorName != "" {
+						cycleErrorJob = jobErrorName
+					}
+				}
+			}
+			return nil
+		},
+	})
+	visited[j.Name] = alreadyVisited
+	return state, cycleErrorJob
+}
+
+func findJobByName(jobName string, jobs atc.JobConfigs) atc.JobConfig {
+	for _, currJob := range jobs {
+		if jobName == currJob.Name {
+			return currJob
+		}
+	}
+	return atc.JobConfig{}
+}
+
+func validateCycle(c atc.Config) error {
+	jobs := c.Jobs
+	cycleExist := false
+	cycleErrorJob := ""
+	emptyMap := make(map[string]int)
+	for _, job := range jobs {
+		cycleExist, cycleErrorJob = detectCycle(job, emptyMap, c)
+		if cycleExist == true {
+			break
```

Comment:
Similarly here, maybe we can just return the error here rather than `break`ing, so we can scope `cycleExist` and `cycleErrorJob` to within the `for` loop (or if you decide to return an `error` from `detectCycle` instead, this just becomes `if err != nil { return err }`)